src/controller.ts

```diff
@@ -1,24 +1,45 @@
+type TextareaKeyboardEventListeners = Partial<{
+  [K in keyof HTMLElementEventMap]: (event: HTMLElementEventMap[K]) => any;
+}>;
+
 /*********************************************
  * Controller for a MathQuill instance
  ********************************************/
+
+type HandlerWithDirectionFunction = NonNullable<
+  HandlerOptions[HandlersWithDirection]
+>;
+type HandlerWithoutDirectionFunction = NonNullable<
+  HandlerOptions[HandlersWithoutDirection]
+>;
+
 class ControllerBase {
   id: number;
   data: ControllerData;
-  root: ControllerRoot;
-  container: $;
+  readonly root: ControllerRoot;
+  readonly container: HTMLElement;
   options: CursorOptions;
   aria: Aria;
   ariaLabel: string;
   ariaPostLabel: string;
-  cursor: Cursor;
+  readonly cursor: Cursor;
   editable: boolean | undefined;
   _ariaAlertTimeout: number;
   KIND_OF_MQ: KIND_OF_MQ;
-  textarea: $ | undefined;
-  textareaSpan: $ | undefined;
-  mathspeakSpan: $ | undefined;
 
-  constructor(root: ControllerRoot, container: $, options: CursorOptions) {
+  textarea: HTMLElement | undefined;
+  private textareaEventListeners: Partial<{
+    [K in keyof HTMLElementEventMap]: (event: HTMLElementEventMap[K]) => any;
+  }> = {};
```

Comment:
Use `Record` instead of creating your own indexed type.

Reply:
I'm not sure how to express this using `Record`.

Reply:
That syntax doesn't appear to work, and neither does
```ts
Record<K in keyof HTMLElementEventMap, (event: HTMLElementEventMap[K]) => any>
```
or
```ts
Record<K extends keyof HTMLElementEventMap, (event: HTMLElementEventMap[K]) => any>
```

TS gives me `Cannot find name 'K'`.

Reply:
Right.  Never mind on this one.